File: typescript/src/workshop/m6/s05_extractfactory/step2/ReservationService.ts
import { IllegalArgumentError, IllegalStateError } from '../../../../shared/errors.js';
import type { Clock } from '../../../shared/time.js';
import type { Reservation } from '../Reservation.js';
import { ReservationFactory } from './ReservationFactory.js';

/**
 * Krok 2: serwis deleguje tworzenie do fabryki; konstruktor bez zmian, więc klienci
 * serwisu niczego nie zauważają.
 */
export class ReservationService {
  private readonly factory: ReservationFactory;
  private readonly takenSeats = new Set<string>();

  constructor(clock: Clock) {
    this.factory = new ReservationFactory(clock);
  }

  reserve(channel: string, email: string, seats: readonly string[]): Reservation {
    this.requireFree(seats);
    const reservation = this.factory.create(channel, email, seats);
    seats.forEach((seat) => this.takenSeats.add(seat));
    return reservation;
  }

  reserveGroup(email: string, seats: readonly string[]): Reservation {
    if (seats.length < 10) {
      throw new IllegalArgumentError('group needs 10+ seats');
    }
    this.requireFree(seats);
    const reservation = this.factory.create('ONLINE', email, seats);
    seats.forEach((seat) => this.takenSeats.add(seat));
    return reservation;
  }

  private requireFree(seats: readonly string[]): void {
    for (const seat of seats) {
      if (this.takenSeats.has(seat)) {
        throw new IllegalStateError(`seat taken: ${seat}`);
      }
    }
  }
}

File: csharp/src/Training.Workshop/M6/S05ExtractFactory/Step1/ReservationService.cs
using Training.Workshop.Shared;

namespace Training.Workshop.M6.S05ExtractFactory.Step1;

/// <summary>
/// Krok 1: Extract Method - tworzenie rezerwacji w jednej metodzie NewReservation.
/// Kolejność zachowana: numer jest pobierany PRZED walidacją kanału (błąd "spala" numer).
/// </summary>
public sealed class ReservationService
{
    private readonly TimeProvider _clock;
    private readonly HashSet<string> _takenSeats = [];
    private int _sequence;

    public ReservationService(TimeProvider clock)
    {
        _clock = clock;
    }

    public Reservation Reserve(string channel, string email, IReadOnlyList<string> seats)
    {
        RequireFree(seats);
        var reservation = NewReservation(channel, email, seats);
        _takenSeats.UnionWith(seats);
        return reservation;
    }

    public Reservation ReserveGroup(string email, IReadOnlyList<string> seats)
    {
        if (seats.Count < 10)
        {
            throw new ArgumentException("group needs 10+ seats");
        }
        RequireFree(seats);
        var reservation = NewReservation("ONLINE", email, seats);
        _takenSeats.UnionWith(seats);
        return reservation;
    }

    private void RequireFree(IReadOnlyList<string> seats)
    {
        foreach (var seat in seats)
        {
            if (_takenSeats.Contains(seat))
            {
                throw new InvalidOperationException("seat taken: " + seat);
            }
        }
    }

    private Reservation NewReservation(string channel, string email, IReadOnlyList<string> seats)
    {
        _sequence++;
        var id = "R" + _sequence;
        if (channel != "ONLINE" && channel != "BOX_OFFICE")
        {
            throw new ArgumentException("unknown channel: " + channel);
        }
        var fee = channel == "ONLINE" ? Money.Of("2.00").Times(seats.Count) : Money.Zero;
        DateTime? expiresAt = channel == "ONLINE"
            ? _clock.GetLocalNow().DateTime.AddMinutes(15)
            : null;
        return new Reservation(id, channel, email, seats.ToList(), fee, expiresAt);
    }
}
